spec/models/defect_spec.rb

```diff
@@ -124,6 +124,20 @@
     end
   end
 
+  describe '#reference_number' do
+    it 'returns a formatted identifier code' do
+      defect = create(:defect, sequence_number: 42)
+      expect(defect.reference_number).to eq('NB-000-042')
+    end
+
+    it 'uses sequential identifiers' do
+      defects = 5.times.map { create(:defect) }
+      ids = defects.map { |d| d.reference_number.scan(/\d/).join('').to_i }
+      first = ids.first
+      expect(ids).to eq((0..4).map { |i| first + i })
```

Comment:
Given that I think we should be starting each Defect unit test with a fresh Defect table I think we might be able to make this part clearer to read too:

```
expect(ids).to eql([1,2,3,4,5])
```

We'd need to be cautious about getting different ids back eg [2,3,4,5,6] if  file wide `before` blocks existed that added defects into the background context for all tests. If this happened, I'd argue that that part should change rather than every test having to defend against the unknowns.

Reply:
I'm not actually getting IDs beginning at 1 in this test but I don't know why. How is the database emptied after each test?

Reply:
We use [a gem called database_cleaner](https://github.com/DatabaseCleaner/database_cleaner).

Reply:
OK, our setup for that looks pretty standard -- transactional cleanup for unit tests, truncation for full-stack things. My hunch is that if a transaction is aborted, any internal counters maintained by the DB will not be rolled back. (If counters were rolled back, you could end up with non-sequential values, and possibly duplicate values, unless you lock the whole table for all queries.) That's why you can end up with gaps in your ID sequence if an `INSERT` fails or is part of a failed transaction. So the counter will not reset between tests.